### PR TITLE
[AIRFLOW-3917] Added config to support running scheduler(with kuberne…

### PR DESCRIPTION
**Pull request by ramandumcs**
…tes executor) outside of kubernetes Cluster

Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-3917\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3917
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [ ] Passes `flake8`

### CODE REVIEW

File: airflow/config_templates/default_airflow.cfg [hunk 1]
```diff
@@ -703,6 +703,11 @@ gcp_service_account_keys =
 # It will raise an exception if called from a process not running in a kubernetes environment.
 in_cluster = True
 
+# Use the cluster_context or config_file to run outside of cluster
+cluster_context =
+config_file =
```
Comment (ashb) [inline airflow/config_templates/default_airflow.cfg]: Are these actually needed? Is setting `in_cluster = False` enough?

File: airflow/contrib/kubernetes/kube_client.py [hunk 2]
```diff
@@ -46,6 +46,8 @@ def _load_kube_config(in_cluster, cluster_context, config_file):
 
 
 def get_kube_client(in_cluster=conf.getboolean('kubernetes', 'in_cluster'),
-                    cluster_context=None,
-                    config_file=None):
+                    cluster_context=None if not conf.get('kubernetes', 'cluster_context') else
+                    conf.get('kubernetes', 'cluster_context'),
+                    config_file=None if not conf.get('kubernetes', 'config_file') else
+                    conf.get('kubernetes', 'config_file')):
```
Comment (ashb) [inline airflow/contrib/kubernetes/kube_client.py]: These defaults are a bit complex to read as they are. Two things:

1. Leave the default param on the function as None (so have `if cluster_context is None:` in the fn body) - this is due to when Python "evaluates" the default values (at import time) which I would like to not depend on having parsed the config.
2. The condition `if not ... else` can be replaced with `conf.get('kubernetes', 'cluster_context', fallback=None)` if you *don't* specify values in the config. I.e `cluster_context =` becomes `# cluster_context =`

**Review by ashb — CHANGES_REQUESTED** [review]

**Comment by ramandumcs** [comment]
RETEST

File: airflow/config_templates/default_airflow.cfg (hunk 1 above)
Comment (ramandumcs) [reply airflow/config_templates/default_airflow.cfg]: Yeah these are required to initialise kube_client if in_cluster is set to False

File: airflow/contrib/kubernetes/kube_client.py (hunk 2 above)
Comment (ramandumcs) [reply airflow/contrib/kubernetes/kube_client.py]: Thanks @ashb I have incorporated comments. I have observed that conf.get('kubernetes', 'cluster_context', fallback=None) fallbacks to None if key i.e cluster_context is not defined in airflow.cfg file. In our case it returns a empty string which breaks the kube_client initialisation

File: airflow/contrib/kubernetes/kube_client.py [hunk 3]
```diff
@@ -48,4 +48,10 @@ def _load_kube_config(in_cluster, cluster_context, config_file):
 def get_kube_client(in_cluster=conf.getboolean('kubernetes', 'in_cluster'),
                     cluster_context=None,
                     config_file=None):
+    if cluster_context is None:
+        if conf.get('kubernetes', 'cluster_context'):
```
Comment (pgagnon) [inline airflow/contrib/kubernetes/kube_client.py]: Merging this if statement with its parent on line 51 would reduce nesting and (in my opinion) improve readability.

File: airflow/contrib/kubernetes/kube_client.py [hunk 4]
```diff
@@ -48,4 +48,10 @@ def _load_kube_config(in_cluster, cluster_context, config_file):
...
+            cluster_context = conf.get('kubernetes', 'cluster_context')
+    if config_file is None:
+        if conf.get('kubernetes', 'config_file'):
```
Comment (pgagnon) [inline airflow/contrib/kubernetes/kube_client.py]: Idem.

File: airflow/contrib/kubernetes/kube_client.py (hunk 3 above)
Comment (ramandumcs) [reply airflow/contrib/kubernetes/kube_client.py]: @pgagnon Thanks. Incorporated comments

File: airflow/contrib/kubernetes/kube_client.py (hunk 2 above)
Comment (ashb) [reply airflow/contrib/kubernetes/kube_client.py]: Which Is why I said

> if you don't specify values in the config. I.e cluster_context = becomes # cluster_context =

That is, make the change in default_airflow.cfg be:


```diff
+# Use the cluster_context or config_file to run outside of cluster
+# cluster_context =
+# config_file =
```

File: airflow/config_templates/default_airflow.cfg (hunk 1 above)
Comment (ashb) [reply airflow/config_templates/default_airflow.cfg]: They aren't though - they are optional. If they aren't provided then the Python Kube library will look at the default config file and use the currently selected context. We should mirror that behaviour.

File: airflow/contrib/kubernetes/kube_client.py [hunk 5]
```diff
@@ -48,4 +48,10 @@ def _load_kube_config(in_cluster, cluster_context, config_file):
 def get_kube_client(in_cluster=conf.getboolean('kubernetes', 'in_cluster'),
                     cluster_context=None,
                     config_file=None):
+    if cluster_context is None \
+       and conf.get('kubernetes', 'cluster_context'):
+        cluster_context = conf.get('kubernetes', 'cluster_context')
```
Comment (ashb) [inline airflow/contrib/kubernetes/kube_client.py]:
```diff
+    if not in_cluster:
+        if cluster_context is None
+            cluster_context = conf.get('kubernetes', 'cluster_context', fallback=None)
+        if config_file is None:
+            config_file = conf.get('kubernetes', 'config_file', fallback=None)
     return _load_kube_config(in_cluster, cluster_context, config_file)
```

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/contrib/kubernetes/kube_client.py (hunk 5 above)
Comment (ramandumcs) [reply airflow/contrib/kubernetes/kube_client.py]: Thanks @ashb. Its my Bad, I misinterpreted your comments. Thanks again for reviewing this.
Incorporated your comments

File: airflow/contrib/executors/kubernetes_executor.py [hunk 6]
```diff
@@ -264,6 +264,14 @@ def _validate(self):
                 'must be set for authentication through user credentials; '
                 'or `git_ssh_key_secret_name` must be set for authentication '
                 'through ssh key, but not both')
+        if conf.getboolean(self.kubernetes_section, 'in_cluster') is False \
+           and not conf.get(self.kubernetes_section, 'config_file') \
+           and not conf.get(self.kubernetes_section, 'cluster_context'):
```
Comment (ashb) [inline airflow/contrib/executors/kubernetes_executor.py]: This change should go now.

File: airflow/config_templates/default_airflow.cfg [hunk 7]
```diff
@@ -703,6 +703,11 @@ gcp_service_account_keys =
 # It will raise an exception if called from a process not running in a kubernetes environment.
 in_cluster = True
 
+# Provide cluster_context or config_file to run it outside kube cluster
```
Comment (ashb) [inline airflow/config_templates/default_airflow.cfg]:
```suggestion
# When running with in_cluster=False change the default cluster_context or config_file 
# options to Kubernetes client. Leave blank these to use default behaviour like `kubectl` has
```

**Review by ashb — CHANGES_REQUESTED** [review]
Final small changes, then this should be good to go!

File: airflow/contrib/executors/kubernetes_executor.py (hunk 6 above)
Comment (ramandumcs) [reply airflow/contrib/executors/kubernetes_executor.py]: @ashb Thanks for helping on this PR. Changes are incorporated

**Review by ashb — APPROVED** [review]

**Pull request merged**